src/sema/contracts.rs

```diff
@@ -691,6 +691,14 @@ fn check_inheritance(contract_no: usize, ns: &mut ast::Namespace) {
         ));
     }
 
+    let contract = &ns.contracts[contract_no];
+    if ns.target.is_substrate() && contract.is_concrete() && !contract.has_public_functions(ns) {
+        diagnostics.push(ast::Diagnostic::warning(
+            contract.loc,
+            format!("Contract '{}' has no public messages. Consider declaring it as an abstract contract instead.", contract.name),
```

Comment:
- lower case contract please
- external or public
-  "messages" is an ink! term, this might be confusing for Solidity developers. I think `public or external functions` makes more sense.
- It might be helpful to explain why this is a problem. Something like "Substrate requires at least one message" or a link to the documentation with a section on this.

Reply:
Are you sure this is the right location for this check?

The `data_locations_01` test contains the follow code:
```solidity
contract foo {
    function bar(uint calldata x) public returns () {
  }
}
```

This erroneously fails with for not having public messages, probably because the previous failure (`data location 'calldata' can only be specified for array, struct or mapping`). I'm having a bit of hard time finding the right place for this check to happen...

Reply:
So if the prototype of a function can't be resolved it's not added to the ast. This is should really be fixed, which is a large issue to solve (lots of code paths to consider)

However, for now, only perform the check if there are no diagnostic errors so far. So `if !ns.diagnostics.any_error() { ... }`

Reply:
> So if the prototype of a function can't be resolved it's not added to the ast. This is should really be fixed, which is a large issue to solve (lots of code paths to consider)

I see, thanks!

> However, for now, only perform the check if there are no diagnostic errors so far. So if !ns.diagnostics.any_error() { ... }

Sorry I forgot to mention, I already tried that and it didn't help in this case